queue.c

```diff
@@ -0,0 +1,95 @@
+#include "queue.h"
+
+status_t
+queue_init (queue_t *queue)
+{
+  queue->head = queue->tail = 0;
+  queue->active = true;
+
+  if (sem_init (&queue->full, 0, 0) != 0)
+    {
+      return S_FAILURE;
+    }
+  if (sem_init (&queue->empty, 0, QUEUE_SIZE) != 0)
+    {
+      return S_FAILURE;
+    }
+
+  if (pthread_mutex_init (&queue->head_mutex, NULL) != 0)
+    {
+      return S_FAILURE;
+    }
+  if (pthread_mutex_init (&queue->tail_mutex, NULL) != 0)
+    {
+      return S_FAILURE;
+    }
+
+  return S_SUCCESS;
+}
+
+status_t
+queue_push (queue_t *queue, task_t *task)
+{
+  if (sem_wait (&queue->empty) != 0)
+    {
+      return S_FAILURE;
```

Comment:
Should set active to false in case of error. Any failed operation puts a queue in an inconsistent state.